File: prep/ucsd-specialisation/algorithms-on-graphs/Week2/topological_sort.rs
use std::collections::{HashMap, HashSet, VecDeque};
use std::io;

fn get_nums() -> (i64, i64) {
    let mut input = String::new();

    io::stdin()
        .read_line(&mut input)
        .expect("failed to read input");

    let nums = input
        .trim()
        .split_whitespace()
        .map(|d| d.parse::<i64>().unwrap())
        .collect::<Vec<_>>();

    (nums[0].clone(), nums[1].clone())
}

struct Vertex {
    vs: HashSet<i64>,
}

impl Vertex {
    fn new() -> Self {
        Vertex { vs: HashSet::new() }
    }

    fn add_edge(&mut self, v: i64) {
        self.vs.insert(v);
    }

    fn get_adjacent_neighbours(&self) -> Vec<i64> {
        let mut ns = Vec::new();

        for v in &self.vs {
            ns.push(*v);
        }
        ns
    }

    fn has_edge(&self, v: i64) -> bool {
        self.vs.contains(&v)
    }
}

struct Graph {
    vertices: Vec<Vertex>,
    size: usize,
}

impl Graph {
    fn new(size: usize) -> Self {
        let mut g = Graph {
            vertices: Vec::new(),
            size: size,
        };
        for _ in 0..size {
            g.vertices.push(Vertex::new());
        }

        g
    }

    fn size(&self) -> usize {
        self.size
    }

    fn add_edge(&mut self, v1: i64, v2: i64) {
        if v1 < 0 || v1 >= self.size as i64 || v2 < 0 || v2 >= self.size as i64 {
            panic!("invalid vertex");
        }

        self.vertices[v1 as usize].add_edge(v2);
    }

    fn get_adjacent_neighbours(&self, v: i64) -> Vec<i64> {
        if v < 0 || v >= self.size as i64 {
            panic!("invalid vertex");
        }

        self.vertices[v as usize].get_adjacent_neighbours()
    }

    fn indegree(&self, v: i64) -> i64 {
        let mut indeg = 0;

        for i in 0..self.size() {
            if self.vertices[i].has_edge(v) {
                indeg += 1;
            }
        }

        indeg
    }
}

fn main() {
    let (n, m) = get_nums();
    let mut g = Graph::new(n as usize);

    for _ in 0..m {
        let (from, to) = get_nums();
        g.add_edge(from - 1, to - 1);
    }

    let mut ordering: Vec<i64> = Vec::new();
    let mut deg = HashMap::new();
    let mut q = VecDeque::new();

    for i in 0..g.size() {
        let d = g.indegree(i as i64);
        deg.insert(i as i64, d);

        if d == 0 {
            q.push_back(i as i64);
        }
    }

    while !q.is_empty() {
        let v = q.pop_front().unwrap();

        ordering.push(v);

        for neighbour in g.get_adjacent_neighbours(v) {
            deg.entry(neighbour).and_modify(|d| *d -= 1);

            if let Some(0) = deg.get(&neighbour) {
                q.push_back(neighbour);
            }
        }
    }

    if ordering.len() != g.size() {
        panic!("cycle detected");
    }

    for v in &ordering {
        print!("{} ", *v + 1);
    }
    println!();
}

File: lc/coreprep/implementation/practice_iterations/main_1/TopologicalSort.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class TopologicalSort {
  static interface Graph {
    void addEdge(int from, int to);
    List<Integer> getAdjacentVertices(int v);
    int indegree(int v);
    int size();
  }

  static class AdjacencySet implements Graph {
    static class Vertex {
      Set<Integer> vs;

      Vertex() { this.vs = new HashSet<>(); }

      boolean hasEdge(int v) { return this.vs.contains(v); }

      void addEdge(int v) { this.vs.add(v); }

      List<Integer> getAdjacentVertices() {
        List<Integer> ns = new ArrayList<>();

        for (int v : this.vs) {
          ns.add(v);
        }

        return ns;
      }
    }

    private List<Vertex> vertices;
    private int size;

    AdjacencySet(int size) {
      this.size = size;
      this.vertices = new ArrayList<>(size);

      for (int i = 0; i < size; i++) {
        this.vertices.add(new Vertex());
      }
    }

    @Override
    public void addEdge(int from, int to) {
      if (from < 0 || from >= this.size || to < 0 || to >= this.size) {
        throw new IllegalArgumentException("invalid vertices");
      }

      this.vertices.get(from).addEdge(to);
    }

    @Override
    public List<Integer> getAdjacentVertices(int v) {
      if (v < 0 || v >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      return this.vertices.get(v).getAdjacentVertices();
    }

    @Override
    public int indegree(int v) {
      if (v < 0 || v >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      int indeg = 0;
      for (int i = 0; i < this.size; i++) {
        if (this.vertices.get(i).hasEdge(v)) {
          indeg++;
        }
      }

      return indeg;
    }

    @Override
    public int size() {
      return this.size;
    }
  }

  // O(E + V) / O(V)
  public static List<Integer> topologicalSort(Graph g) {
    Queue<Integer> q = new ArrayDeque<>();

    Map<Integer, Integer> indeg = new HashMap<>();
    for (int i = 0; i < g.size(); i++) {
      int id = g.indegree(i);
      if (id == 0) {
        q.add(i);
      }
      indeg.put(i, id);
    }

    List<Integer> ordering = new ArrayList<>();

    while (!q.isEmpty()) {
      int v = q.poll();

      ordering.add(v);

      for (int neighbour : g.getAdjacentVertices(v)) {
        indeg.put(neighbour, indeg.get(neighbour) - 1);

        if (indeg.get(neighbour) == 0) {
          q.add(neighbour);
        }
      }
    }

    if (ordering.size() != g.size()) {
      throw new IllegalStateException("cycle detected");
    }

    return ordering;
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      Graph g = new AdjacencySet(n);

      int m = in.nextInt();
      for (int i = 0; i < m; i++) {
        int from = in.nextInt();
        int to = in.nextInt();

        g.addEdge(from, to);
      }

      var ordering = topologicalSort(g);
      for (int v : ordering) {
        System.out.printf("%d ", v);
      }
      System.out.println();
    }
  }
}